sound/soc/sof/pcm.c

```diff
@@ -213,7 +215,34 @@ static int sof_pcm_hw_params(struct snd_soc_component *component,
 
 	dev_dbg(component->dev, "stream_tag %d", pcm.params.stream_tag);
 
-	/* send IPC to the DSP */
+	/* if this is a repeated hw_params without hw_free, skip setting up widgets */
+	if (spcm->stream[substream->stream].list)
+		goto send_ipc;
+
+	/* first set up connected DAPM widgets */
+	for_each_rtd_cpu_dais(rtd, j, dai) {
+		struct snd_soc_dapm_widget_list *list;
+		int paths = snd_soc_dapm_dai_get_connected_widgets(dai, substream->stream, &list,
+								   dpcm_end_walk_at_be);
+
+		if (paths < 0) {
+			dev_warn(component->dev, "error: dai %s has no valid %s path\n", dai->name,
+				 substream->stream == SNDRV_PCM_STREAM_PLAYBACK ?
+				 "playback" : "capture");
+			return paths;
+		}
+
+		spcm->stream[substream->stream].list = list;
```

Comment:
shouldn't this assignment only be done in case the below call succeeds? Or you could keep it like this, but also remove the `list` parameter from the call to `sof_widget_list_setup()` below and from `sof_widget_list_free()` and assign `NULL` inside `sof_widget_list_setup()` in case of failure.

Reply:
done